bookkeeper-server/src/test/java/org/apache/bookkeeper/meta/ZkLedgerUnderreplicationManagerTest.java

```diff
@@ -0,0 +1,244 @@
+     */
+    @Test
+    public void testMarkLedgerUnderreplicatedBasic() throws Exception {
+        long ledgerId = 0xabcdef;
+        Collection<String> missingBookies = Lists.newArrayList("bookie-1");
+
+        long prevCtime = -1L;
+
+        // mark when it hasn't been marked before
+        FutureUtils.result(urMgr.markLedgerUnderreplicatedAsync(ledgerId, missingBookies));
+        UnderreplicatedLedgerFormat urLedgerFormat = urMgr.getLedgerUnreplicationInfo(ledgerId);
+        assertEquals(missingBookies, urLedgerFormat.getReplicaList());
+        assertTrue(urLedgerFormat.getCtime() > prevCtime);
+        prevCtime = urLedgerFormat.getCtime();
+
+        // mark when it has been marked. but the missing bookies already duplicated there
+        FutureUtils.result(urMgr.markLedgerUnderreplicatedAsync(ledgerId, missingBookies));
+        urLedgerFormat = urMgr.getLedgerUnreplicationInfo(ledgerId);
+        assertEquals(missingBookies, urLedgerFormat.getReplicaList());
+        assertTrue(urLedgerFormat.getCtime() >= prevCtime);
+        prevCtime = urLedgerFormat.getCtime();
+
+        // mark with new bookies when it has been marked
+        Collection<String> newMissingBookies = Lists.newArrayList("bookie-2", "bookie-3");
+        FutureUtils.result(urMgr.markLedgerUnderreplicatedAsync(ledgerId, newMissingBookies));
+        urLedgerFormat = urMgr.getLedgerUnreplicationInfo(ledgerId);
+        assertEquals(
+            Lists.newArrayList("bookie-1", "bookie-2", "bookie-3"),
+            urLedgerFormat.getReplicaList()
+        );
+        assertTrue(urLedgerFormat.getCtime() >= prevCtime);
+    }
+
+    /**
+     * Test {@link ZkLedgerUnderreplicationManager#markLedgerUnderreplicatedAsync(long, Collection)} handles corrupted
+     * data.
+     */
+    @Test
+    public void testMarkLedgerWithCorruptedDataExists() throws Exception {
+        long ledgerId = 0xabcdee;
+        String ledgerPath = urMgr.getUrLedgerZnode(ledgerId);
+        ZkUtils.createFullPathOptimistic(
+            zk, ledgerPath, "junk data".getBytes(UTF_8), ZkUtils.getACLs(baseServerConf), CreateMode.PERSISTENT);
+        Collection<String> missingBookies = Lists.newArrayList("bookie-1");
+        try {
+            FutureUtils.result(urMgr.markLedgerUnderreplicatedAsync(ledgerId, missingBookies));
+            fail("Should fail to mark ledger underreplicated if there is already corrupted data on zookeeper");
+        } catch (ReplicationException re) {
+            assertTrue(re.getMessage().contains("Invalid underreplicated ledger data for ledger " + ledgerPath));
+        }
+        byte[] data = zk.getData(ledgerPath, null, null);
+        assertEquals("junk data", new String(data, UTF_8));
+    }
+
+    @Test
+    public void testMarkLedgerUnderreplicatedConcurrently() throws Exception {
+        final int numLedgers = 20;
+        List<CompletableFuture<Void>> futures = Lists.newArrayListWithExpectedSize(numLedgers);
```

Comment:
this is soon to be deprecated method - https://google.github.io/guava/releases/22.0/api/docs/com/google/common/collect/Lists.html#newArrayListWithExpectedSize-int-

Reply:
new ArrayList<>(numLedgers() ?